From: Takeuchi <https://github.com/taksan>
Subject: [PATCH 0005] add named multi-instance support

Source: https://github.com/waydroid/waydroid/pull/1990
Backported to Waydroid 1.6.3 for miodroid.

Adds --instance/-I, with per-instance work directories, D-Bus names,
container names, network interfaces, and binderfs mount points.

diff --git a/data/scripts/waydroid-net.sh b/data/scripts/waydroid-net.sh
--- a/data/scripts/waydroid-net.sh	2026-05-29 06:02:24.000000000 +1000
+++ b/data/scripts/waydroid-net.sh	2026-10-06 16:09:20.375740948 +1100
@@ -1,26 +1,40 @@
 #!/bin/sh -
 
 varrun="/run/waydroid-lxc"
+WORK="/var/lib/waydroid"
+iface_name="waydroid0"
+
+WAYDROID_INSTANCE=$2
+if [ ! -z "$WAYDROID_INSTANCE" ]; then
+    varrun="/run/waydroid-${WAYDROID_INSTANCE}-lxc"
+    WORK="/var/lib/waydroid.${WAYDROID_INSTANCE}"
+    iface_name="waydroid-${WAYDROID_INSTANCE}"
+fi
 varlib="/var/lib"
 net_link_key="lxc.net.0.link"
 case "$(lxc-info --version)" in [012].*) net_link_key="lxc.network.link" ;; esac
-vnic=$(awk "\$1 == \"$net_link_key\" {print \$3}" /var/lib/waydroid/lxc/waydroid/config)
-: ${vnic:=waydroid0}
+vnic=$(awk "\$1 == \"$net_link_key\" {print \$3}" $WORK/lxc/waydroid/config)
+: ${vnic:=$iface_name}
 
-if [ "$vnic" != "waydroid0" ]; then
+if [ "$vnic" != "$iface_name" ]; then
     echo "vnic is $vnic, bailing out"
     exit 0
 else 
-    echo "vnic is waydroid0"
+    echo "vnic is $iface_name"
 fi
 
+MD5HASH=$(echo -n "$WAYDROID_INSTANCE" | md5sum | cut -d ' ' -f1 )
+AS_DECIMAL=$(printf "%d" 0x$(echo "$MD5HASH" | cut -c1-8))
+SUB_ADDR=$(($AS_DECIMAL % 16 + 2))
+NET_ID=$(($AS_DECIMAL % 16 + 241))
+
 USE_LXC_BRIDGE="true"
-LXC_BRIDGE="${vnic}"
+LXC_BRIDGE="${iface_name}"
 LXC_BRIDGE_MAC="00:16:3e:00:00:01"
-LXC_ADDR="192.168.240.1"
+LXC_ADDR="192.168.${NET_ID}.${SUB_ADDR}"
 LXC_NETMASK="255.255.255.0"
-LXC_NETWORK="192.168.240.0/24"
-LXC_DHCP_RANGE="192.168.240.2,192.168.240.254"
+LXC_NETWORK="192.168.${NET_ID}.0/24"
+LXC_DHCP_RANGE="192.168.${NET_ID}.16,192.168.${NET_ID}.254"
 LXC_DHCP_MAX="253"
 LXC_DHCP_CONFILE=""
 LXC_DHCP_PING="true"
diff --git a/tools/actions/container_manager.py b/tools/actions/container_manager.py
--- a/tools/actions/container_manager.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/actions/container_manager.py	2026-10-06 16:09:20.376740944 +1100
@@ -163,7 +163,8 @@
 
     # Networking
     command = [tools.config.tools_src +
-               "/data/scripts/waydroid-net.sh", "start"]
+               "/data/scripts/waydroid-net.sh", "start", args.instance]
+
     tools.helpers.run.user(args, command)
 
     # Sensors
@@ -236,7 +237,7 @@
 
         # Networking
         command = [tools.config.tools_src +
-                   "/data/scripts/waydroid-net.sh", "stop"]
+                   "/data/scripts/waydroid-net.sh", "stop", args.instance]
         tools.helpers.run.user(args, command, check=False)
 
         #TODO: remove NFC hacks
diff --git a/tools/actions/initializer.py b/tools/actions/initializer.py
--- a/tools/actions/initializer.py	2026-10-06 16:09:20.344741085 +1100
+++ b/tools/actions/initializer.py	2026-10-06 16:09:20.376740944 +1100
@@ -186,6 +186,7 @@
     helpers.lxc.set_lxc_config(args)
     helpers.lxc.make_base_props(args)
     if status != "STOPPED":
+        helpers.dbus.setup_policy(args)
         try:
             if "running_init_in_service" in args:
                 tools.actions.container_manager.do_start(args, session)
diff --git a/tools/actions/session_manager.py b/tools/actions/session_manager.py
--- a/tools/actions/session_manager.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/actions/session_manager.py	2026-10-06 16:09:20.376740944 +1100
@@ -4,6 +4,7 @@
 import os
 import signal
 import sys
+from tools.helpers import instance
 import tools.config
 import tools.helpers.ipc
 from tools import services
@@ -39,7 +40,7 @@
 
 def start(args, unlocked_cb=None, background=True):
     try:
-        _name = dbus.service.BusName("id.waydro.Session", dbus.SessionBus(), do_not_queue=True)
+        _name = dbus.service.BusName(instance.get_session_dbus_name(), dbus.SessionBus(), do_not_queue=True)
     except dbus.exceptions.NameExistsException:
         logging.error("Session is already running")
         if unlocked_cb:
diff --git a/tools/config/__init__.py b/tools/config/__init__.py
--- a/tools/config/__init__.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/config/__init__.py	2026-10-06 16:09:20.376740944 +1100
@@ -68,7 +68,8 @@
     "lcd_density": "0",
     "background_start": "true"
 }
-session_defaults["waydroid_user_state"] = session_defaults["xdg_data_home"] + "/waydroid"
+instance_suffix = f".{os.environ['MIODROID_INSTANCE']}" if os.environ.get("MIODROID_INSTANCE") else ""
+session_defaults["waydroid_user_state"] = session_defaults["xdg_data_home"] + f"/waydroid{instance_suffix}"
 session_defaults["waydroid_data"] = session_defaults["waydroid_user_state"] + "/data"
 if session_defaults["pulse_runtime_path"] == "None":
     session_defaults["pulse_runtime_path"] = session_defaults["xdg_runtime_dir"] + "/pulse"
diff --git a/tools/helpers/arguments.py b/tools/helpers/arguments.py
--- a/tools/helpers/arguments.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/helpers/arguments.py	2026-10-06 16:09:20.376740944 +1100
@@ -1,6 +1,7 @@
 # Copyright 2021 Oliver Smith
 # SPDX-License-Identifier: GPL-3.0-or-later
 import argparse
+import re
 
 try:
     import argcomplete
@@ -146,6 +147,9 @@
     # Logging
     parser.add_argument("-l", "--log", dest="log", default=None,
                         help="path to log file")
+    parser.add_argument("-I", "--instance", type=validate_instance_name,
+                     help="instance name (default is nothing)")
+
     parser.add_argument("--details-to-stdout", dest="details_to_stdout",
                         help="print details (e.g. build output) to stdout,"
                              " instead of writing to the log",
@@ -180,3 +184,8 @@
     # Parse and extend arguments (also backup unmodified result from argparse)
     args = parser.parse_args()
     return args
+
+def validate_instance_name(name):
+    if not re.match("^[a-zA-Z0-9_]+$", name):
+        raise argparse.ArgumentError(None, "instance name must consist of letters, numbers and underscores only")
+    return name
diff --git a/tools/helpers/drivers.py b/tools/helpers/drivers.py
--- a/tools/helpers/drivers.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/helpers/drivers.py	2026-10-06 16:09:25.346719066 +1100
@@ -8,8 +8,11 @@
 import struct
 import tools.config
 import tools.helpers.run
+from tools.helpers.instance import get_binderfs_dir, wrap_suffix
 from contextlib import suppress
 
+def wrap_instance_suffix(input_array):
+    return [wrap_suffix(input) for input in input_array]
 
 BINDER_DRIVERS = [
     "anbox-binder",
@@ -17,18 +20,18 @@
     "bonder",
     "binder"
 ]
-VNDBINDER_DRIVERS = [
+VNDBINDER_DRIVERS = wrap_instance_suffix([
     "anbox-vndbinder",
     "vndpuddlejumper",
     "vndbonder",
     "vndbinder"
-]
-HWBINDER_DRIVERS = [
+])
+HWBINDER_DRIVERS = wrap_instance_suffix([
     "anbox-hwbinder",
     "hwpuddlejumper",
     "hwbonder",
     "hwbinder"
-]
+])
 
 
 def isBinderfsLoaded(args):
@@ -58,7 +61,7 @@
         return IOC(READ|WRITE, _type, nr, size)
 
     BINDER_CTL_ADD = IOWR(98, 1, 264)
-    with open('/dev/binderfs/binder-control', 'rb') as f:
+    with open(f'{get_binderfs_dir()}/binder-control', 'rb') as f:
         for node in binder_dev_nodes:
             node_struct = struct.pack('256sII', bytes(node, 'utf-8'), 0, 0)
             with suppress(FileExistsError):
@@ -96,13 +99,13 @@
                 logging.error(output.strip())
 
         if isBinderfsLoaded(args):
-            command = ["mkdir", "-p", "/dev/binderfs"]
+            command = ["mkdir", "-p", get_binderfs_dir()]
             tools.helpers.run.user(args, command, check=False)
-            command = ["mount", "-t", "binder", "binder", "/dev/binderfs"]
+            command = ["mount", "-t", "binder", "binder", get_binderfs_dir()]
             tools.helpers.run.user(args, command, check=False)
             allocBinderNodes(args, binder_dev_nodes)
             command = ["ln", "-s"]
-            command.extend(glob.glob("/dev/binderfs/*"))
+            command.extend(glob.glob(f"{get_binderfs_dir()}/*"))
             command.append("/dev/")
             tools.helpers.run.user(args, command, check=False)
 
diff --git a/tools/helpers/__init__.py b/tools/helpers/__init__.py
--- a/tools/helpers/__init__.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/helpers/__init__.py	2026-10-06 16:09:20.376740944 +1100
@@ -12,4 +12,6 @@
 import tools.helpers.gpu
 import tools.helpers.protocol
 import tools.helpers.version
+import tools.helpers.instance
+import tools.helpers.dbus
 import tools.helpers.logging
diff --git a/tools/helpers/instance.py b/tools/helpers/instance.py
--- a/tools/helpers/instance.py	1970-01-01 10:00:00.000000000 +1000
+++ b/tools/helpers/instance.py	2026-10-06 16:09:20.376740944 +1100
@@ -0,0 +1,35 @@
+from tools.helpers.arguments import arguments
+
+# hack: we pass version because tools.config depends on this file and that
+# would cause a circular dependency. We just need the instance name here.
+INSTANCE_NAME = arguments("").instance
+
+def get_container_dbus_name():
+    return f"id.waydro.Container{get_suffix()}"
+
+def get_container_name():
+    return "waydroid"
+
+def get_session_dbus_name():
+    return f"id.waydro.Session{get_suffix()}"
+
+def get_work_dir():
+    return f"/var/lib/waydroid{get_suffix()}"
+
+def get_data_dir():
+    return f"/waydroid{get_suffix()}/data"
+
+def get_inet_name():
+    return f"waydroid{get_suffix_dash()}"
+
+def get_suffix():
+    return f".{INSTANCE_NAME}" if INSTANCE_NAME else ""
+
+def get_suffix_dash():
+    return f"-{INSTANCE_NAME}" if INSTANCE_NAME else ""
+
+def get_binderfs_dir():
+    return f"/dev/binderfs{get_suffix_dash()}"
+
+def wrap_suffix(input):
+    return f"{input}{get_suffix_dash()}"
diff --git a/tools/helpers/ipc.py b/tools/helpers/ipc.py
--- a/tools/helpers/ipc.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/helpers/ipc.py	2026-10-06 16:09:20.377740939 +1100
@@ -4,9 +4,10 @@
 # Currently implemented as FIFO
 import os
 import dbus
+from . import instance
 
 def DBusContainerService(object_path="/ContainerManager", intf="id.waydro.ContainerManager"):
-    return dbus.Interface(dbus.SystemBus().get_object("id.waydro.Container", object_path), intf)
+    return dbus.Interface(dbus.SystemBus().get_object(instance.get_container_dbus_name(), object_path), intf)
 
 def DBusSessionService(object_path="/SessionManager", intf="id.waydro.SessionManager"):
-    return dbus.Interface(dbus.SessionBus().get_object("id.waydro.Session", object_path), intf)
+    return dbus.Interface(dbus.SessionBus().get_object(instance.get_session_dbus_name(), object_path), intf)
diff --git a/tools/helpers/lxc.py b/tools/helpers/lxc.py
--- a/tools/helpers/lxc.py	2026-05-29 06:02:24.000000000 +1000
+++ b/tools/helpers/lxc.py	2026-10-06 16:09:20.377740939 +1100
@@ -12,6 +12,7 @@
 import platform
 import gbinder
 import tools.config
+from tools.helpers.instance import get_container_name, get_inet_name, get_work_dir
 import tools.helpers.run
 from contextlib import suppress
 from pathlib import Path
@@ -160,7 +161,20 @@
 
     command = ["mkdir", "-p", lxc_path]
     tools.helpers.run.user(args, command)
-    command = ["sh", "-c", "cat {} > \"{}\"".format(' '.join('"{0}"'.format(w) for w in config_snippets), lxc_path + "/config")]
+
+    if os.path.exists(lxc_path + "/config"):
+        os.remove(lxc_path + "/config")
+    for w in config_snippets:
+        with open(w, "r") as input:
+            lines = input.readlines()
+
+        with open(lxc_path + "/config", "a") as output:
+            for line in lines:
+                output_line = line.replace("/var/lib/waydroid", get_work_dir())
+                output_line = output_line.replace("waydroid0", get_inet_name())
+                output.write(output_line)
+    
+    tools.helpers.run.user(args, command)
     tools.helpers.run.user(args, command)
     command = ["sed", "-i", "s/LXCARCH/{}/".format(platform.machine()), lxc_path + "/config"]
     tools.helpers.run.user(args, command)
@@ -378,7 +392,7 @@
         shutil.copy(filename, tools.config.defaults["host_perms"])
 
 def status(args):
-    command = ["lxc-info", "-P", tools.config.defaults["lxc"], "-n", "waydroid", "-sH"]
+    command = ["lxc-info", "-P", tools.config.defaults["lxc"], "-n", get_container_name(), "-sH"]
     try:
         return tools.helpers.run.user(args, command, output_return=True).strip()
     except Exception:
@@ -399,7 +413,7 @@
 
 def start(args):
     command = ["lxc-start", "-P", tools.config.defaults["lxc"],
-               "-F", "-n", "waydroid", "--", "/init"]
+               "-F", "-n", get_container_name(), "--", "/init"]
     tools.helpers.run.user(args, command, output="background")
     wait_for_running(args)
     # Workaround lxc-start changing stdout/stderr permissions to 700
@@ -408,11 +422,11 @@
 
 def stop(args):
     command = ["lxc-stop", "-P",
-               tools.config.defaults["lxc"], "-n", "waydroid", "-k"]
+               tools.config.defaults["lxc"], "-n", get_container_name(), "-k"]
     tools.helpers.run.user(args, command)
 
 def freeze(args):
-    command = ["lxc-freeze", "-P", tools.config.defaults["lxc"], "-n", "waydroid"]
+    command = ["lxc-freeze", "-P", tools.config.defaults["lxc"], "-n", get_container_name()]
     tools.helpers.run.user(args, command)
 
 def unfreeze(args):
